modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java

```diff
@@ -242,21 +245,44 @@ ScriptScope compile(Loader loader, String name, String source, CompilerSettings
      * @return The bytes for compilation.
      */
     byte[] compile(String name, String source, CompilerSettings settings, Printer debugStream) {
+        return compile(name, source, settings, debugStream, null, null, null);
+    }
+
+    /**
+     * Runs the two-pass compiler to generate a Painless script with option visitors for each major phase.
+     */
+    byte[] compile(String name, String source, CompilerSettings settings, Printer debugStream,
+                   UserTreeVisitor<ScriptScope> semanticPhaseVisitor,
+                   UserTreeVisitor<ScriptScope> irPhaseVisitor,
+                   IRTreeVisitor<WriteScope> asmPhaseVisitor) {
         String scriptName = Location.computeSourceName(name);
         ScriptClassInfo scriptClassInfo = new ScriptClassInfo(painlessLookup, scriptClass);
         SClass root = Walker.buildPainlessTree(scriptName, source, settings);
         ScriptScope scriptScope = new ScriptScope(painlessLookup, settings, scriptClassInfo, scriptName, source, root.getIdentifier() + 1);
+
         new PainlessSemanticHeaderPhase().visitClass(root, scriptScope);
         new PainlessSemanticAnalysisPhase().visitClass(root, scriptScope);
-        // TODO: Make this phase optional #60156
+        if (semanticPhaseVisitor != null) {
+            semanticPhaseVisitor.visitClass(root, scriptScope);
+        }
+
         new DocFieldsPhase().visitClass(root, scriptScope);
         new PainlessUserTreeToIRTreePhase().visitClass(root, scriptScope);
+        if (irPhaseVisitor != null) {
+            irPhaseVisitor.visitClass(root, scriptScope);
+        }
+
         ClassNode classNode = (ClassNode)scriptScope.getDecoration(root, IRNodeDecoration.class).getIRNode();
         new DefaultStringConcatenationOptimizationPhase().visitClass(classNode, null);
         new DefaultConstantFoldingOptimizationPhase().visitClass(classNode, null);
         new DefaultStaticConstantExtractionPhase().visitClass(classNode, scriptScope);
         classNode.setDebugStream(debugStream);
-        new DefaultIRTreeToASMBytesPhase().visitScript(classNode);
+
+        WriteScope writeScope = WriteScope.newScriptScope();
```

Comment:
I'm not sure this change is an improvement. None of our phases quite align correctly as many require the ScriptScope. I would consider the WriteScope more equivalent to the SemanticScope hidden as part of the SemanticPhase. This should probably be redesigned into a high level script scope that gets passed in at the root of all phases and then each individual phase should add their own scoping as necessary.

Reply:
I pulled this out to be able to have access to the writeScope, as this change does not implement the `IRTreeVisitor`, I can revert this change if desired.

Reply:
Created a new method.